### PR TITLE
Remove useless copy on zip file load

### PR DESCRIPTION
**Pull request by driazati**
Instead of copying to a buffer, then setting a tensor's storage with that buffer, create a storage directly from the file

Differential Revision: [D21889537](https://our.internmc.facebook.com/intern/diff/21889537/)

### CODE REVIEW

File: torch/csrc/jit/python/init.cpp [hunk 1]
```diff
@@ -694,6 +694,29 @@ void initJITBindings(PyObject* module) {
...
           })
+      .def(
+          "get_storage_from_record",
+          [](PyTorchStreamReader& self,
+             const std::string& key,
+             size_t numel,
+             py::object dd) {
+            at::DataPtr data;
+            size_t size;
+            std::tie(data, size) = self.getRecord(key);
```
Comment (voznesenskym) [inline torch/csrc/jit/python/init.cpp:705]: `auto [data, size] = self.getRecord(key);`

File: torch/serialization.py [hunk 2]
```diff
@@ -486,7 +486,6 @@ def persistent_id(obj):
         if storage.device.type == 'cpu':
             # If it's on the CPU we can directly copy it into the zip file
             num_bytes = storage.size() * storage.element_size()
-            buf = io.BytesIO()
```
Comment (voznesenskym) [inline torch/serialization.py:489]: Was this just spurious/unused?

File: torch/csrc/jit/python/init.cpp [hunk 3]
```diff
@@ -694,6 +694,29 @@ void initJITBindings(PyObject* module) {
...
+            auto scalar_type =
+                reinterpret_cast<THPDtype*>(dd.ptr())->scalar_type;
```
Comment (voznesenskym) [inline torch/csrc/jit/python/init.cpp:707]: nit: Can this be cast with a safer cast? static_cast perhaps? what type is `dd.ptr()` before cast?

File: torch/csrc/jit/python/init.cpp (hunk 1 above)
Comment (voznesenskym) [reply torch/csrc/jit/python/init.cpp:705]: Actually, you aren't using size? Maybe just do 

`auto data = self.getRecord(key).get(0); ?`

File: torch/csrc/jit/python/init.cpp [hunk 4]
```diff
@@ -694,6 +694,29 @@ void initJITBindings(PyObject* module) {
             std::tie(data, size) = self.getRecord(key);
             return py::bytes(reinterpret_cast<const char*>(data.get()), size);
           })
+      .def(
+          "get_storage_from_record",
+          [](PyTorchStreamReader& self,
+             const std::string& key,
+             size_t numel,
+             py::object dd) {
```
Comment (voznesenskym) [inline torch/csrc/jit/python/init.cpp:702]: nit - why numel and dd? I see `py::object` usually called obj or buffer in other places.

File: torch/csrc/jit/python/init.cpp [hunk 5]
```diff
@@ -694,6 +694,29 @@ void initJITBindings(PyObject* module) {
...
+            std::tie(data, size) = self.getRecord(key);
+            auto scalar_type =
+                reinterpret_cast<THPDtype*>(dd.ptr())->scalar_type;
+
+            auto storage = c10::Storage(
+                at::CPU(scalar_type).typeMeta(),
+                numel,
+                std::move(data),
+                /*allocator=*/nullptr,
+                /*resizable=*/false);
```
Comment (voznesenskym) [inline torch/csrc/jit/python/init.cpp:714]: If these are reasonable and safe defaults, should we make them default args?

**Review by voznesenskym — APPROVED** [review]

**Comment by voznesenskym** [comment]
LGTM predicated on passing CI

**Comment by malfet** [comment]
After the merge all the builds are broken with:
```
ay 22 02:13:50 /var/lib/jenkins/workspace/torch/csrc/jit/python/init.cpp:742:28: error: no matching constructor for initialization of 'c10::Storage'
May 22 02:13:50             auto storage = c10::Storage(
May 22 02:13:50                            ^
May 22 02:13:50 /var/lib/jenkins/workspace/c10/core/Storage.h:29:3: note: candidate constructor not viable: requires 5 arguments, but 6 were provided
May 22 02:13:50   Storage(
May 22 02:13:50   ^
May 22 02:13:50 /var/lib/jenkins/workspace/c10/core/Storage.h:15:3: note: candidate constructor not viable: requires 4 arguments, but 6 were provided
May 22 02:13:50   Storage(
May 22 02:13:50   ^
May 22 02:13:50 /var/lib/jenkins/workspace/c10/core/Storage.h:12:3: note: candidate constructor not viable: requires single argument 'ptr', but 6 arguments were provided
May 22 02:13:50   Storage(c10::intrusive_ptr<StorageImpl> ptr) : storage_impl_(std::move(ptr)) {}
May 22 02:13:50   ^
May 22 02:13:50 /var/lib/jenkins/workspace/c10/core/Storage.h:7:16: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 6 were provided
May 22 02:13:50 struct C10_API Storage {
May 22 02:13:50                ^
May 22 02:13:50 /var/lib/jenkins/workspace/c10/core/Storage.h:7:16: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 6 were provided
May 22 02:13:50 /var/lib/jenkins/workspace/c10/core/Storage.h:11:3: note: candidate constructor not viable: requires 0 arguments, but 6 were provided
May 22 02:13:50   Storage() {}
May 22 02:13:50   ^
```
For example see https://app.circleci.com/pipelines/github/pytorch/pytorch/172958/workflows/efd8bbfa-085d-437f-9443-119e77d38580/jobs/5556802/steps

**Comment by voznesenskym** [comment]
@malfet  that's not good... let me see if I can fix it.

cc @driazati

**Comment by malfet** [comment]
@voznesenskym I've reverted the diff, so please rebase against top of tree, fix compilation issues and try to reland it

**Comment by voznesenskym** [comment]
It's not my diff, unfortunately, I am just following along as other work I have dovetails with it nicely (https://github.com/pytorch/pytorch/pull/38379/) and I reviewed it to speed things up. Stamp was predicated on CI passing ;)

**Comment by voznesenskym** [comment]
@driazati please take a look!

**Pull request closed without merging**